# Dag 7 - ochtend


### Wat ik vandaag ga doen:
- Importeren Library's
- Laad de dataset 'iris'
- Maak een feature matrix X en een target vector y
- reshape de target vector y
- splits de data in train / test data
- maak een eenvoudig neuraal netwerk
- Compileer en bekijk het neuraal netwerk
- Train het model
- evaluate het model

In deze opgave classificeren afbeeldingen van bloemen uit de iris dataset met behulp van een neuraal netwerk.

In [11]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder

Laad de dataset en maak een feature matrix X en een target vector y.

In [12]:
iris = sns.load_dataset('iris')
# we gebruiken .values om een Numpy array te krijgen in plaats van een Pandas DataFrame
x_iris = iris.drop('species', axis=1).values 
y_iris = iris['species'].values

print(iris)

     sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]


>__x_iris__ = de sepal_length totaan de petal_width. Deze features zijn belangrijk bij het uitzoeken welke 'Species' het is. Dit is de feature matrix. Omdat het een matrix is van features (wow).


>__y_iris__ = de daadwerkelijke species, aangezien dit namen (strings) zijn moeten we deze nog verranderen zodat het neuraal netwerk ze kan lezen. Dit is de target vector. Omdat dit een lijst is met targets die het neurale netwerk moet gaan raken.

>__conclusie__ = Dus we kunnen hieruit zeggen dat we op de x-as de features hebben en op de y-as de waardes die we willen dat het neuraal netwerk vindt (targets). 

In [13]:
y_iris = y_iris.reshape(-1, 1) # maak een array van array's, dit is nodig voor de volgende stap

Een neuraal netwerk kan niet omgaan met categorische labels zoals de namen van de planten, maar heeft getallen als uitput nodig. We hebben hier drie klassen (setosa, versicolor, virginica) daarom maken we een target vector met per label drie waarden. Een 1 voor de eerste waarde correspondeert met 'setosa', een 1 voor de 2e waade met 'versicolor' en een 1 voor de derde waarde met 'virginica'.

Doe dit met behulp van het `OneHotEncoder` object uit sklearn.

In [14]:
encoder = OneHotEncoder(sparse=False) 
y_encoded = encoder.fit_transform(y_iris)
print(y_encoded)

[[1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0.

>__conclusie__: Aangezien de computer niet weet wat een string is, en als output getallen geeft. Moeten we onze waardes omschrijven in getallen. Dit kan erg makkelijk met ***'OneHotEncoder'***. Dit encode onze data zodat het neuraal netwerk het snapt

We hebben de data nu in een geschikt formaat. Splits de data in training en test data.

In [16]:
X_train, X_test, y_train, y_test = train_test_split(x_iris, y_encoded, test_size=0.2, random_state = 6)

> __conclusie__: Nu zijn we op het punt aangekomen om onze dataset te ***splitsen*** in een train en test set. Dit doen we om overfitting te voorkomen. We geven 0.2 (20%) van onze totale dataset mee om uiteindelijk te gaan testen. In dit geval is dat 150 * 0.2 = 30 rows.

Maak een eenvoudig neuraal netwerk met 3 lagen, waarin elke neuron in een laag met elk neuron in de vorige laag verbonden is ('Dense'). 
Voeg voor de eeste laag een parameter `input_shape` toe die aangeeft hoeveel features een element in 'X' heeft.
Het aantal neuronen in de laaste laag is gelijk aan het aantal waarden in een label in `y`. 

In [11]:
from keras.models import Sequential
from keras.layers import Dense

model = Sequential()
model.add(Dense(4, input_shape = (4,), activation = 'relu'))
model.add(Dense(3, activation = 'relu'))
model.add(Dense(3, activation = 'softmax'))

> __conlusie__: We hebben hier een "eenvoudig" neuraal netwerk gemaakt. Dit kan met de library keras. Nu gebruiken we een model die gebasseerd is op ***Sequential*** lagen. En 3 lagen toegevoegd met ***model.add***. De Dense is de dikte van de laag (hoeveel neuronen). Daarnaast hebben we een softmax functie meegegeven aan de laastse laag om zo de output makkelijk weer te geven omdat we moeten alleen heoven te kijken naar welk het dichst bij de 1 zit.

>**LET OP!** De laaste laag moet altijd hetzelfde aantal neuronen (Dense) hebben als je hoeveel uitkomsten die je krijgt.

Compileer en bekijk het netwerk door onderstaande code te runnen.

In [12]:
from tensorflow.keras.optimizers import Adam

model.compile(Adam(learning_rate=0.01),'categorical_crossentropy',metrics=['accuracy'])

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 4)                 20        
_________________________________________________________________
dense_1 (Dense)              (None, 3)                 15        
_________________________________________________________________
dense_2 (Dense)              (None, 3)                 12        
Total params: 47
Trainable params: 47
Non-trainable params: 0
_________________________________________________________________


> __conclusie__: Nu hebben we het netwerk gecompileerd en gebruiken categorical crossentropy als een functie omdat die erg goed is in ___classification___. We meten met de metric genaamd ___accuracy___. Soms ook wel acc genoemd.

Nu kunnen we ons model trainen met behulp van de `fit` methode.

In [13]:
model.fit(X_train, y_train, epochs = 25, batch_size = 5) 

Epoch 1/25
24/24 [==============================] - 0s 566us/step - loss: 1.3079 - accuracy: 0.3583
Epoch 2/25
24/24 [==============================] - 0s 522us/step - loss: 1.1091 - accuracy: 0.3917
Epoch 3/25
24/24 [==============================] - 0s 566us/step - loss: 1.0971 - accuracy: 0.3583
Epoch 4/25
24/24 [==============================] - 0s 566us/step - loss: 1.0709 - accuracy: 0.4917
Epoch 5/25
24/24 [==============================] - 0s 522us/step - loss: 1.0097 - accuracy: 0.3583
Epoch 6/25
24/24 [==============================] - 0s 522us/step - loss: 0.8681 - accuracy: 0.5583
Epoch 7/25
24/24 [==============================] - 0s 566us/step - loss: 0.7480 - accuracy: 0.6917
Epoch 8/25
24/24 [==============================] - 0s 566us/step - loss: 0.6694 - accuracy: 0.7917
Epoch 9/25
24/24 [==============================] - 0s 566us/step - loss: 0.6053 - accuracy: 0.7500
Epoch 10/25
24/24 [==============================] - 0s 522us/step - loss: 0.5571 - accuracy: 0.8500

> __conclusie__: We hebben nu voor het eerst getrained (wow). De Epochs is 1 volledige run door het neurale netwerk, en batch size is na hoeveel runs hij zichzelf leert. Daarnaast zie je loss en accuracy staan. Hoe dichter de accuracy bij 1 zit, hoe ebter. En hoe kleiner de loss is, hoe beter.

Met `model_evaluate` kunnen we bepalen hoe goed het model werkt op de test data.

In [14]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print(test_acc)

1/1 [==============================] - 0s 121ms/step - loss: 0.1190 - accuracy: 0.9667
0.9666666388511658


>__conclusie__: Zoals hierboven vermeld wordt model_evaluate gebruikt om te bepalen hoe goed het model werkt op de test data.

Experimenteer met bovenstaand model. Wat gebeurt er als je meer lagen toevoegt of een laag weghaalt of het aantal neuronen in een laag verandert? 

Probeer ook eens andere activatiefuncties dan `relu`, zoals `tanh` of `sigmoid`.

Welke invloed heeft het aantal epochs?

In [15]:
model.fit(X_train, y_train, epochs = 50, batch_size = 5) # 2x zoveel epochs

Epoch 1/50
24/24 [==============================] - 0s 566us/step - loss: 0.1228 - accuracy: 0.9750
Epoch 2/50
24/24 [==============================] - 0s 522us/step - loss: 0.1304 - accuracy: 0.9750
Epoch 3/50
24/24 [==============================] - 0s 522us/step - loss: 0.1300 - accuracy: 0.9750
Epoch 4/50
24/24 [==============================] - 0s 566us/step - loss: 0.1033 - accuracy: 0.9750
Epoch 5/50
24/24 [==============================] - 0s 566us/step - loss: 0.1782 - accuracy: 0.9250
Epoch 6/50
24/24 [==============================] - 0s 609us/step - loss: 0.1467 - accuracy: 0.9417
Epoch 7/50
24/24 [==============================] - 0s 566us/step - loss: 0.1089 - accuracy: 0.9667
Epoch 8/50
24/24 [==============================] - 0s 566us/step - loss: 0.1008 - accuracy: 0.9750
Epoch 9/50
24/24 [==============================] - 0s 566us/step - loss: 0.1057 - accuracy: 0.9833
Epoch 10/50
24/24 [==============================] - 0s 522us/step - loss: 0.1312 - accuracy: 0.9583

In [16]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print(test_acc) # Een accuracy van 0.9666... Hij is nu erg geoptimaliseerd voor deze dataset.

1/1 [==============================] - 0s 13ms/step - loss: 0.0837 - accuracy: 0.9667
0.9666666388511658


We kunnen dezelfde aanpak gebruiken voor andere data, bijvoorbeeld voor het herkennen van cijfers. We gebruiken de `MNist` dataset die bestaat uit plaatjes van cijfers. Ieder plaatje bestaat uit 28 x 28 pixels.

In [17]:
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

import warnings
warnings.filterwarnings('ignore')

In [18]:
train_images.shape

(60000, 28, 28)

In [19]:
len(train_labels)

60000

In [20]:
print(train_labels)

[5 0 4 ... 5 6 8]


Een eenvoudig neuraal netwerk heeft een array van inputwaarden nodig i.p.v. een 2D afbeelding.
We maken van de inputs (waarden tussen 0 en 255) getallen tussen 0 en 1.

In [21]:
train_images = train_images.reshape((60000, 28 * 28)) # lijst van waarden i.p.v. 2D afbeelding
train_images = train_images.astype('float32') / 255 # getallen tussen 0 en 1

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

Voor elk label (een cijfer) maken we een array met 10 waarden: 9 nullen en een één, waarbij de positie van de één aangeeft om welk cijfer het gaat (One hot encoding). Dit keer gebruiken we hiervoor de `keras` functie `to_categorical`.

In [22]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

print(train_labels)

ImportError: cannot import name 'to_categorical' from 'keras.utils' (E:\Anaconda3\lib\site-packages\keras\utils\__init__.py)

Maak nu een neuraal netwerk met twee lagen, een eerste laag met 512 neuronen en een output laag met 10 neuronen. Geef bij de eerste laag aan hoeveel inputs er zijn en wat de activatiefunctie is. De output laag heeft een `softmax` activatiefunctie.

compileer het netwerk.

train het netwerk 5 epochs met een batch_size van 128.

Evalueer het netwerk op de test set.

Experimenteer met verschillende netwerken. Varieer het aantal lagen, het aantal neuronen, de activatiefunties en het aantal epochs.